### DISCUSSION TITLE
Using specific fixed model input size

### DISCUSSION THREAD

**Discussion by vladimirmujagic** [openpifpaf/openpifpaf#475] [Q&A]
Hello,

I am currently trying to test resnet50 model using different input sizes:
```
640x640
640x480
320x320
320x240
160x160
160x120
```
Code responsible for resizing of original image:
```
    def preprocess_factory(args):
        """ Utility method to instantiate preprocessor """

        preprocess = [transforms.NormalizeAnnotations()]
        if args.long_edge:
            preprocess.append(transforms.RescaleAbsolute(args.long_edge))

        preprocess.append(transforms.CenterPad(args.long_edge))

        # if args.batch_size > 1:
        #     assert args.long_edge, '--long-edge must be provided for batch size > 1'
        #     preprocess.append(transforms.CenterPad(args.long_edge))
        # else:
        #     preprocess.append(transforms.CenterPadTight(16))

        return transforms.Compose(preprocess + [transforms.EVAL_TRANSFORM])
```
So for square inputs adding `preprocess.append(transforms.CenterPad(args.long_edge))` instead of commented block allowed me to test with square input sizes.

The question is how can I resize other dimension to fit to my desired value because having variable input sizes is not possible using tensorrt inference engine for instance. Every image has to be resized to some specified `HxW`

I would like to stick to original code as much as possible instead of using my own cv resize function to keep aspect ratio

**Comment by svenkreiss** [comment]
Thanks for your report. OpenPifPaf doesn't have a tensorrt backend option. The `export_onnx` file has already option to specify fixed width and height. Which tool do you need improved?

**Comment by vladimirmujagic** [comment]
Thank you for quick response,

So to be more precise, I am trying to reuse your code (preprocessing, postprocessing) and use tensorrt inference instead of torch inference, also testing with onnxruntime. As mentioned for square input sizes this process is working and I can reproduce original results when doing evaluation.

But what i cannot do with original preprocessing code is to resize to target `WxH` it can either work with square input or it adds 16 pixels to other dimension (not being long edge) - square inputs are created when using batch size > 1.

I can solve this by using my custom function to preprocess the input image to target `WxH` and preserve aspect ratio.

But i am trying to evade modifying original code as much as possible and stay within the framework (your original code) for the purpose of replicating results.

Essentially can your transforms somehow be used to preprocess image to target `WxH` for instance `160x120`

I am aware that you can use `export_onnx` to generate any input size `WxH` but then you need preprocessing code to resize image to that `WxH` and i would like to use original code to do that. The reason why I am doing this is to try and align 1-1 with your implementation and reproduce same evaluation results across different engines.

With square inputs i was able to prove `torch == onnxruntime == tensorrt` for input sizes `129x129; 321x321; 641x641`. Would like to test this with non square inputs

Additional note i am using `dev` branch commit: `6385c33b2c32e305d045362e662d732a9dd3c78a`

Regards,
Vladimir

**Comment by vladimirmujagic** [comment]
To add some more info

I am using `OpenCV` library to implement preprocessing code, main purpose of this function is to resize image to target `WxH` and preserve aspect ratio by padding either left, right or top, bottom based on image aspect `W>H` or `H>W`.

This function is useful to me because using your preprocessing code I cannot resize to target `WxH` I can only specify long edge or get a square input size.

Here is the code:
```
def _resize_and_pad_keep_aspect_ratio(img, size, pad_color=0):
    h, w = img.shape[:2]
    sh, sw = size

    if h > sh or w > sw:
        interp = cv2.INTER_AREA
    else:
        interp = cv2.INTER_CUBIC

    aspect = w / h
    if aspect > 1:
        new_w = sw
        new_h = np.round(new_w / aspect).astype(int)
        pad_vert = abs((sh - new_h)) / 2
        pad_top, pad_bot = np.floor(pad_vert).astype(int), np.ceil(pad_vert).astype(int)
        print('horizontal', pad_top, pad_bot)
        pad_left, pad_right = 0, 0
    elif aspect < 1:
        new_h = sh
        new_w = np.round(new_h * aspect).astype(int)
        pad_horz = abs((sw - new_w)) / 2
        pad_left, pad_right = np.floor(pad_horz).astype(int), np.ceil(pad_horz).astype(int)
        print('vertical', pad_right, pad_left)
        pad_top, pad_bot = 0, 0
    else:
        new_h, new_w = sh, sw
        pad_left, pad_right, pad_top, pad_bot = 0, 0, 0, 0

    if len(img.shape) is 3 and not isinstance(pad_color, (list, tuple, np.ndarray)):
        pad_color = [pad_color] * 3

    scaled_img = cv2.resize(img, (new_w, new_h), interpolation=interp)
    scaled_img = cv2.copyMakeBorder(
        scaled_img, pad_top, pad_bot,
        pad_left, pad_right,
        borderType=cv2.BORDER_CONSTANT,
        value=pad_color
    )

    return scaled_img, new_w, new_h, pad_left, pad_top
```

Would this kind of preprocessing code be compatible with original code meaning there won't be a huge delta when running evaluation?. I will of course run tests to check but generally does it make sense to preprocess images in this way when using openpifpaf?

**Comment by vladimirmujagic** [comment]
Here are the test results:

Using default implementation (branch: `dev`, commit: `6385c33b2c32e305d045362e662d732a9dd3c78a`)
Images are preprocessed t o `641x641`
```
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.683
Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.884
Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.747
Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.656
Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.731
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.730
Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.910
Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.788
Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.685
Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.795
```

Using `OpenCV` function posted earlier and custom `annotation_inverse` implementation based on preprocessing code. Images are preprocessed to `641x641`
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.602
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.827
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.659
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.565
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.664
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.649
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.851
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.702
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.592
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.729
```
Remapping code:
```
for annotation in post_processed_output:
    annotation.data[:, 0] -= self.pad_left
    annotation.data[:, 1] -= self.pad_top
    
    annotation.data[:, 0] = annotation.data[:, 0] * (self.or_w / self.new_w)
    annotation.data[:, 1] = annotation.data[:, 1] * (self.or_h / self.new_h)
```
Where `self.pad_left`, `self.pad_top`, `self.or_w`, `self.or_h` and `new_w` and `new_h` are obtained from  `_resize_and_pad_keep_aspect_ratio` function described in previous post
